File: src/main/scala/com/jackywong/scala/parsing/my/Parser.scala
package com.jackywong.scala.parsing.my

import com.jackywong.scala.parsing.my.Helpers.~

/**
  * Created by huangziqi on 2019/7/1
  * 每一个Parser仅仅就是一个需要ParseState并且返回Result的函数而已
  */
trait Parser[+A] {
  def exec(state:ParseState): ParseResult[A]

  def label(msg: String): Parser[A]
  def scope(msg: String): Parser[A]
  def attempt: Parser[A]
  def slice: Parser[String]

  def flatMap[B](f: A => Parser[B]): Parser[B]
  def map[B](f: A => B): Parser[B]
  def map2[B,C](p2: => Parser[B])(f: (A,B) => C): Parser[C]
  def product[B](p2: => Parser[B]): Parser[(A,B)]
  def **[B](p2: => Parser[B]): Parser[(A, B)] = product(p2)
  def productWith[B](p2: => Parser[B]): Parser[A~B]
  def ~[B](p2: => Parser[B]): Parser[A~B] = productWith(p2)
  def as[B](b: B): Parser[B]

  //解析的逻辑运算，组合子
  //且 this & that
  def and[B>:A](p2: => Parser[B]): Parser[B]
  def &[B>:A](p2: => Parser[B]): Parser[B] = and(p2)
  //或 this | that
  def or[B>:A](p2: => Parser[B]): Parser[B]
  def |[B>:A](p2: => Parser[B]): Parser[B] = or(p2)
  //省略自身 this *> that
  def skipL[B](p2: => Parser[B]): Parser[B]
  def *>[B](p2: => Parser[B]): Parser[B] = skipL(p2)
  //省略下一个 this <* that
  def skipR(p2: => Parser[_]): Parser[A]
  def <*(p2: => Parser[_]): Parser[A] = skipR(p2)
  //获取option
  def opt: Parser[Option[A]]
  //以p2分割，重复出现零次或多次
  def sep(p2: Parser[_]): Parser[List[A]]
  //以p2分割，重复出现一次或多次
  def sep1(p2: Parser[_]): Parser[List[A]]
  //重复N次
  def listOfN(n: Int): Parser[List[A]]
}

File: src/main/java/com/jackywong/parsing/Parser.java
package com.jackywong.parsing;

import com.jackywong.safer.Tuple;
import com.jackywong.scala.parsing.my.ParseResult;
import com.jackywong.scala.parsing.my.ParseState;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * Created by huangziqi on 2019/7/16
 */
public interface Parser<A> {
    ParseResult<? extends A> exec(ParseState state);

    Parser<? extends A> label(String msg);
    Parser<? extends A> scope(String msg);
    Parser<? extends A> attempt();
    Parser<String> slice();

    <B> Parser<B> flatMap(Function<A,Parser<B>> f);
    <B> Parser<B> map(Function<A,B> f);
    <B,C> Parser<C> map2(Supplier<Parser<B>> p2, Function<Tuple<A,B>,C> f);
    <B> Parser<Tuple<A,B>> product(Supplier<Parser<B>> p2);
    <B> Parser<B> as(B b);

    Parser<? super A> and(Supplier<Parser<?>> p2);
    Parser<? super A> or(Supplier<Parser<?>> p2);
    <B> Parser<B> skipL(Supplier<Parser<B>> p2);
    Parser<A> skipR(Supplier<Parser<?>> p2);
    Parser<Optional<A>> opt();
    Parser<List<A>> sep(Parser<?> p2);
    Parser<List<A>> sep1(Parser<?> p2);
    Parser<List<A>> listOfN(int n);
}
